include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,317 @@
+              zFrac0 = shifted.x;
+              zFrac1 = shifted.y;
+              zFrac2 = shifted.z;
+              zExp = 0;
+              
+              if (roundNearestEven)
+              {
+                 increment = zFrac2 < 0u;
+              }
+              else
+              {
+                 if (zSign != 0u)
+                 {
+                    //increment = (FLOAT_ROUNDING_MODE == FLOAT_ROUND_DOWN) && (zFrac2 != 0u);
+                 }
+                 else
+                 {
+                    //increment = (FLOAT_ROUNDING_MODE == FLOAT_ROUND_UP) && (zFrac2 != 0u);
+                 }
+              }
+           }
+        
+           if (increment)
+           {
+              nbl::hlsl::uint32_t2 added = add64(zFrac0, zFrac1, 0u, 1u);
+              zFrac0 = added.x;
+              zFrac1 = added.y;
+              zFrac1 &= ~((zFrac2 + uint32_t(zFrac2 == 0u)) & uint32_t(roundNearestEven));
+           }
+           else
+           {
+              zExp = nbl::hlsl::lerp(zExp, 0, (zFrac0 | zFrac1) == 0u);
+           }
+           
+           return packFloat64(zSign, zExp, zFrac0, zFrac1);
+        }
+
+
+    }
+
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+
+        storage_t data;
+
+        // constructors
+        // TODO: specializations
+        template <typename T>
+        static emulated_float64_t create(T val)
+        { 
+            emulated_float64_t output;
+            output.data = val;
+            return output;
+        }
+        
+        template<>
+        static emulated_float64_t create(double val)
+        {
+            emulated_float64_t output;
+            output.data = reinterpret_cast<uint64_t&>(val);
+            return output;
+        }
+
+        // arithmetic operators
+        emulated_float64_t operator+(const emulated_float64_t rhs)
+        {
+            emulated_float64_t retval;
+            retval.data = data + rhs.data;
+            return retval;
+        }
+
+        emulated_float64_t operator-(const emulated_float64_t rhs)
+        {
+            emulated_float64_t retval;
+            retval.data = data - rhs.data;
+            return retval;
+        }
+
+        emulated_float64_t operator*(const emulated_float64_t rhs)
+        {
+            emulated_float64_t retval = emulated_float64_t::create(0u);
+            
+            
+            uint32_t lhsLow = uint32_t(data & 0x00000000FFFFFFFFull);
+            uint32_t rhsLow = uint32_t(rhs.data & 0x00000000FFFFFFFFull);
+            uint32_t lhsHigh = uint32_t((data & 0x000FFFFF00000000ull) >> 32);
+            uint32_t rhsHigh = uint32_t((rhs.data & 0x000FFFFF00000000ull) >> 32);
+            uint32_t lhsExp = uint32_t((data >> 52) & 0x7FFull);
+            uint32_t rhsExp = uint32_t((rhs.data >> 52) & 0x7FFull);
+
+            int32_t exp = int32_t(lhsExp + rhsExp) - 0x400u;

```

Comment:
you can do this math in 16bit int if you want even

Reply:
no need i guess, unless there is something i don't know and using int16_t may be beneficial?